### ISSUE TITLE
Python 3.10 rebuild

### ISSUE THREAD

**Issue by oldherl** [archlinuxcn/repo#2590]
### 问题类型 / Type of issues

* 其它 / other

### 受影响的软件包 / Affected packages

* [x] bypy-git (@petronny)
* [x] clang-git (@solaraquarion)
* [x] conan (@heavysink)
* [x] isso (@PeterCxy)
* [x] lldb-git (@solaraquarion)
* [x] llvm-git (@solaraquarion)
* [x] mirror (@ideal)
* [x] music-dl (@dctxmei)
* [x] pip-tools (@qwIvan)
* [x] python-albumentations (@hubutui)
* [x] python-django-formtools (@yuyichao)
* [x] python-lvis-git (@hubutui)
* [x] python-misaka (@PeterCxy)
* [x] python-node-semver (@heavysink)
* [x] python-pyqt4 (@heavysink)   3ca38d8
* [x] python-torchstat (@hubutui)
* [x] retext-git (@farseerfc)
* [x] syncthing-gtk-python3 (@SilverRainZ)
* [x] youtube-dl-anime-git (@heavysink)

----

These packages contain files for python 3.9, and should be updated to python 3.10 now. It's a similar issue as #2001.

Most packages have been rebuilt by lilac, but the above are missing.

For packages failed with its dependencies not rebuilt successfully, triggering a rebuild is needed once the dependencies are fixed.

For other packages, please investigate.

Please only tick the checkbox only AFTER the package is actually BUILT successfully, not just checking in the script change.

**Comment by yuyichao** [comment]
> python-django-formtools

This one failed to build and I was not able to reproduce it locally yet. Something pip related. @lilydjwg did the behavior of `pypi_pre_build` changed recently?

**Comment by oldherl** [comment]
If "alias: python" is missing in the "update_on" section in your lilac.yaml, please add it.

**Comment by yuyichao** [comment]
llvm-git had a bug in python version detection and is not fixed. `clang-git`, `lldb-git`, `llvm-git` should be fixed after the next build.

**Comment by lilydjwg** [comment]
@yuyichao No. Either setuptools changes, or version 2.3 changes. You need to specify dependencies `python-django` and `python-setuptools-scm` or setuptools will try to use pip to install them.

**Comment by cubercsl** [comment]
package base [`pyqt4`](https://github.com/archlinuxcn/repo/tree/master/archlinuxcn/pyqt4)  rebuild failed since python 3.9.

https://build.archlinuxcn.org/packages/#/pyqt4

**Comment by yuyichao** [comment]
And `python-django-formtools` should be fixed now.

**Comment by SolarAquarion** [comment]
@yuyichao so there's a problem with the upstream? or is there something i need to change.

**Comment by oldherl** [comment]
> package base [`pyqt4`](https://github.com/archlinuxcn/repo/tree/master/archlinuxcn/pyqt4) rebuild failed since python 3.9.
> 
> https://build.archlinuxcn.org/packages/#/pyqt4

Good catch. I'm adding the python-pyqt4 package to the list. But I would suggest to just drop that package.

**Comment by yuyichao** [comment]
@SolarAquarion not upstream, problem with the pig build but I should have fixed it.

**Comment by hubutui** [comment]
orphaning `python-torchstat`  https://github.com/archlinuxcn/repo/issues/2576

**Comment by yuyichao** [comment]
> Please only tick the checkbox only AFTER the package is actually BUILT successfully, not just checking in the script change.

If the package build failed it's not particularly useful to track such issues here unless it's because the upstream doesn't support python3.10. It's also not like we are the official repo with some ways to release all packages at once and have some way to block the python upgrade if some critical packages doesn't work.

Moreover, if the build failed, the maintainer WILL get a notification to start investigating the issue and can come back to uncheck the box. However, that doesn't happen if the build succeeds causing some of the packages to remain unchecked long after they are actually fixed (it happened every time on this kind of issues before). I'll be much more OK with such a policy (checking only after build) if either,

1. It doesn't take hours to see if the build actually works, or
2. There can be a notification sent out when such a build does work, or
3. The check list can be filled automatically.

Otherwise, I'm afraid the list will just remain very out of date as it always has been. I came back by accident this time (forgot to close a tab and somehow notices that some checkmarks are missing) and I would have never checked my packages otherwise.

**Comment by xz-dev** [comment]
proxmark3-iceman also need rebuild

**Comment by lilydjwg** [comment]
> There can be a notification sent out when such a build does work, or

You can call the `notify_maintainers` function to send out a mail. You can condition it by date or version.

**Comment by oldherl** [comment]
Python 3.10 rebuild isn't like common out-of-date issue. If a package doesn't update to python 3.10, it's unusable to an Arch user. Therefore maintainers should make reasonable effort to catch up the python 3.10 update in a timely manner.

> If the package build failed it's not particularly useful to track such issues here unless it's because the upstream doesn't support python3.10 (although even in that case it might be more useful to track it separately/elsewhere) It's also not like we are the official repo with some ways to release all packages at once and have some way to block the python upgrade if some critical packages doesn't work.

Correct. We can't block the core/python package from distributing. But there's still some actions an archlinuxcn maintainer can take.
* Trigger the rebuild.
* If that fails, debug the PKGBUILD and fix the build.
* If a trivial fix won't do, communicate with upstream to work out a fix.
* If that still doesn't work, (e.g. the upstream doesn't have interest in catching up new python version), the package should be dropped.

>  I'm afraid the list will just remain very out of date as it always has been. I came back by accident this time (forgot to close a tab and somehow notices that some checkmarks are missing) and I would have never checked my packages otherwise.
This list doesn't have to be minute-updated. If you make a change in the evening, you can check the package and tick the box next evening. This list consists of only 19 packages. Each maintainer only has no more than 4 packages to check. I don't think it's too much for them to check. In addition, I am going to check the remaining packages and tick them by myself every day till 2021-12-31 or when this issue is closed.

**Comment by oldherl** [comment]
> proxmark3-iceman also need rebuild

Are you sure it doesn't work with python3.10? There isn't any files in /usr/lib/python3.9/site-packages/ in that package.

**Comment by yuyichao** [comment]
> You can call the notify_maintainers function in post_build to send out a mail. You can condition it by date or version.

It should be a trigger by commit message, or I'd have to edit it every time I want a version to notify me and left it in there.

> Python 3.10 rebuild isn't like common out-of-date issue. If a package doesn't update to python 3.10, it's unusable to an Arch user. Therefore maintainers should make reasonable effort to catch up the python 3.10 update in a timely manner.

But it is similar to common broken package for random reason issues.

> Each maintainer only has no more than 4 packages to check. I don't think it's too much for them to check.

As a matter of fact, the reason I prefer automatic build and notification is exactly because I don't have to manually check periodically. And even though I was trying to do the check-after-build thing before, I have inevitably left fixed packages uncheck and never remember to come back to fix them.

> In addition, I am going to check the remaining packages and tick them by myself every day till 2021-12-31 or when this issue is closed.

This certainly works for me. It has never (or at least often not) been the case before (status has been out of date for weeks on previous issues like this) which is why I decided to check them and get failed package notification if they did fail this time. If you are checking, then I'll just be skipping the check for this kind of issues in the future (and just keep my own list that get checked when the build script is fixed).

**Comment by yuyichao** [comment]
> proxmark3-iceman also need rebuild

Probably doesn't apply to this, but there may be precompiled python file in non-standard path. If checking for file content is too much, checking for `.cpython-39.` in the file names should at least be doable. (it would have caught `python-seafdav`, which I was in the process of fixing, this way).

**Comment by oldherl** [comment]
> If you are checking, then I'll just be skipping the check for this kind of issues in the future (and just keep my own list that get checked when the build script is fixed).

I'm checking this list, but no promise to any future lists.

**Comment by xz-dev** [comment]
> > proxmark3-iceman also need rebuild
> 
> Are you sure it doesn't work with python3.10? There isn't any files in /usr/lib/python3.9/site-packages/ in that package.

Its not a python package, but Im sure the error is because of python version.

**Comment by oldherl** [comment]
> > > proxmark3-iceman also need rebuild
> > 
> > 
> > Are you sure it doesn't work with python3.10? There isn't any files in /usr/lib/python3.9/site-packages/ in that package.
> 
> Its not a python package, but Im sure the error is because of python version.

Where is the error message?

**Comment by heavysink** [comment]
Removed pyqt4. The other two packages by me have some problems in PKGBUILD and I fixed them. Will be rebuild automatically by lilac if everything is good.

**Comment by ideal** [comment]
Updated in upstream, https://aur.archlinux.org/packages/mirror
Will it be rebuilt automatically?

**Comment by oldherl** [comment]
> Updated in upstream, https://aur.archlinux.org/packages/mirror Will it be rebuilt automatically?

Should be.

**Comment by oldherl** [comment]
All those packages are updated to Python 3.10 now. Closing.
